### PR TITLE
Agent-Client TLS

### LINKED ISSUE #2299: Encrypted agent connection
Add an option to talk with the agent using TLS instead of plain TCP

### PR DESCRIPTION
**Pull request by Razz4780**
Closes #2299 
(Only relevant when running with operator)

1. `AgentManagement::create_connection` now returns plain IO types (can wrapped in TLS)  
2. Agent uses Mozilla's roots from `webpki_roots` to initialize a TLS connection.
3. Some other minor changes were made to agent initialization - improved tracing, removed blocking on network IO from the main loop, changed client id to be incremental (we're not going to have more than u32 agent clients, c'mon)

### CODE REVIEW

File: mirrord/kube/src/api/container.rs [hunk 1]
```diff
@@ -35,6 +35,7 @@ pub struct ContainerParams {
     pub name: String,
     pub gid: u16,
     pub port: u16,
+    pub extra_env: Vec<(String, String)>,
```
Comment (meowjesty) [inline mirrord/kube/src/api/container.rs:38]: :memo: 

Why is this here, how is this used? Docs please.

File: mirrord/kube/src/api/kubernetes.rs [hunk 2]
```diff
@@ -91,89 +74,93 @@ impl KubernetesAPI {
...
-                "mirrord has detected that you might be running on a cluster with a \
-                 service mesh and `network.incoming.mode = \"mirror\"`, which is currently \
-                 unsupported. You can set `network.incoming.mode` to \"steal\" (check out the\
-                 `http_filter` configuration value if you only want to steal some of the traffic).",
-            );
-        }
-        Ok(())
-    }
+#[cfg(not(feature = "incluster"))]
+pub trait UnpinStream:
```
Comment (meowjesty) [inline mirrord/kube/src/api/kubernetes.rs:80]: :memo: 

Docs on this helper trait.

Also, what is the deal with the `#cfg` for this? Docs for that please.

Comment (meowjesty) [reply mirrord/kube/src/api/kubernetes.rs:80]: :memo: 

`impl<T> UnpinStream for T where` could you also doc who `T` is, please?

File: mirrord/kube/src/api/kubernetes.rs [hunk 3]
```diff
@@ -184,57 +171,55 @@ impl AgentManagment for KubernetesAPI {
...
-        Ok(wrap_raw_connection(
-            port_forwarder
-                .take_stream(connect_info.agent_port)
-                .ok_or(KubeApiError::PortForwardFailed)?,
-        ))
+        let stream = port_forwarder
+            .take_stream(connect_info.agent_port)
+            .ok_or(KubeApiError::PortForwardFailed)?;
+
+        let stream: Box<dyn UnpinStream> = Box::new(stream);
```
Comment (meowjesty) [inline mirrord/kube/src/api/kubernetes.rs:178]: Do we really need this to be generic, such that we want a `dyn trait` here?

File: mirrord/protocol/src/lib.rs [hunk 4]
```diff
@@ -88,3 +88,12 @@ impl FromStr for MeshVendor {
         }
     }
 }
+
+/// Name of environment variable that can be used to instruct the agent to use TLS when acception
```
Comment (meowjesty) [inline mirrord/protocol/src/lib.rs:92]:
```suggestion
/// Name of environment variable that can be used to instruct the agent to use TLS when accepting
```

File: mirrord/agent/src/client_connection.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+/// Wrapper over client's network connection with the agent.
+pub struct ClientConnection {
+    framed: ConnectionFramed,
+    client_id: ClientId,
+}
+
+impl ClientConnection {
+    /// Wraps the given [`TcpStream`] into this struct.
+    /// If a [`TlsConnector`] is given, it is used to first build a make a TLS connection using the
```
Comment (meowjesty) [inline mirrord/agent/src/client_connection.rs:19]:
```suggestion
    /// If a [`TlsConnector`] is given, it is used to first build and make a TLS connection using the
```

**Review by meowjesty — CHANGES_REQUESTED** [review]
Some initial notes :notes:

File: mirrord/agent/src/main.rs [hunk 6]
```diff
@@ -66,27 +69,42 @@ mod outgoing;
...
-#[derive(Debug)]
+/// Keeps track of next client id.
+/// Stores common data used when serving client connections.
+/// Can be cheaply cloned and passed to per-client background tasks.
+#[derive(Clone)]
 struct State {
-    clients: HashSet<ClientId>,
-    index_allocator: IndexAllocator<ClientId, 100>,
+    /// [`ClientId`] for the next client that connects to this agent.
+    next_client_id: Arc<AtomicU32>,
```
Comment (meowjesty) [inline mirrord/agent/src/main.rs:86]: Was this the last place we were using the `IndexAllocator`?

File: mirrord/agent/src/main.rs [hunk 7]
```diff
@@ -66,27 +69,42 @@ mod outgoing;
...
+    tls_connector: Option<TlsConnector>,
 }
 
 impl State {
     /// Return [`Err`] if container runtime operations failed.
     pub async fn new(args: &Args, watch: drain::Watch) -> Result<State> {
+        let tls_connector = args.use_tls.then(|| {
+            TlsConnector::from(Arc::new(
+                ClientConfig::builder()
+                    .dangerous()
```
Comment (meowjesty) [inline mirrord/agent/src/main.rs:102]: :skull_and_crossbones: 

This is why I wish rust allowed emojis as identifiers, just imagine, _fn_ :skull_and_crossbones:.

File: mirrord/agent/src/main.rs [hunk 8]
```diff
@@ -659,37 +628,33 @@ async fn start_agent(args: Args, watch: drain::Watch) -> Result<()> {
...
-                    bg_tasks.clone(),
-                    cancellation_token.clone(),
-                    args.base_protocol_version.clone(),
-                )
-                .await?
-            {
-                clients.push(client)
-            };
+            trace!(peer = %addr, "start_agent -> First connection accepted");
+            clients.spawn(state.clone().serve_client_connection(
```
Comment (meowjesty) [inline mirrord/agent/src/main.rs:642]: Ah, I see now why you want to `clone` state, instead of making a new one on each connection.

File: mirrord/agent/src/main.rs (hunk 6 above)
Comment (Razz4780) [reply mirrord/agent/src/main.rs:85]: Nope, agent's `FileManager`, `TcpConnectionSniffer` and task for outgoing UDP still use it

File: mirrord/agent/src/main.rs (hunk 7 above)
Comment (Razz4780) [reply mirrord/agent/src/main.rs:101]: :skull_and_crossbones: **danger** :skull_and_crossbones:

File: mirrord/kube/src/api/kubernetes.rs (hunk 2 above)
Comment (Razz4780) [reply mirrord/kube/src/api/kubernetes.rs:82]: Added a doc for the `incluster` feature (crate level)

File: mirrord/kube/src/api/container.rs (hunk 1 above)
Comment (meowjesty) [reply mirrord/kube/src/api/container.rs:38]: You mean this `pub const AGENT_TLS_ENV: &str = "MIRRORD_AGENT_USE_TLS";` right?

Comment (Razz4780) [reply mirrord/kube/src/api/container.rs:38]: Yeah, previously this env was meant to be passed in the `extra_env` field. Now it's handled in this crate, based on `use_tls`

**Review by meowjesty — APPROVED** [review]
:+1:
